### PR TITLE
Enter insert mode if an input inside a shadow DOM is focused

### PR DESCRIPTION
**Pull request by mrmr1993**
This fixes issue #853. Testbed [here](http://jsfiddle.net/0xL12q5w/).

### CODE REVIEW

File: content_scripts/mode_insert.coffee [hunk 1]
```diff
@@ -53,6 +56,21 @@ class InsertMode extends Mode
...
+        else if event.target.shadowRoot
+          # A focusable element inside the shadow DOM might have been selected. If so, we can catch the focus
+          # event inside the shadow DOM. This fixes #853.
+          shadowRoot = event.target.shadowRoot
+          eventListeners = {}
+          for type in ["focus", "blur"]
+            eventListeners[type] = do (type) ->
+              (event) ->
+                handlerStack.bubbleEvent type, event
+                if type == "blur"
```
Comment (smblott-github) [inline content_scripts/mode_insert.coffee]: @mrmr1993.  What if there are two or more inputs inside the shadow DOM, and one is first focused, then another (and perhaps then another too)?  Wouldn't the `blur` events cause these listeners to be removed before the shadow DOM itself has lost the focus?  Might b40647c48ae7d8016aa77c418eed42c042000f46 be better?

**Comment by smblott-github** [comment]
Ping.
(Since you seem to be active, @mrmr1993.  And I'll get to the Settings PR soon.)

File: content_scripts/mode_insert.coffee (hunk 1 above)
Comment (mrmr1993) [inline content_scripts/mode_insert.coffee]: Yes, b40647c48ae7d8016aa77c418eed42c042000f46 looks like a better alternative. I have no computer this week, so I haven't tested it.

Comment (smblott-github) [inline content_scripts/mode_insert.coffee]: OK.  This is a nice improvement.  I'll merge it, then cherry pick b40647c48ae7d8016aa77c418eed42c042000f46.

**Pull request merged**